src/v/datalake/record_translator.cc

```diff
@@ -55,17 +55,132 @@ std::optional<size_t> get_redpanda_idx(const iceberg::struct_type& val_type) {
+        system_data->fields.emplace_back(std::move(headers_list));
+    }
+
+    system_data->fields.emplace_back(iceberg::binary_value{std::move(key)});
+    return system_data;
+}
+
 } // namespace
 
 std::ostream& operator<<(std::ostream& o, const record_translator::errc& e) {
     switch (e) {
     case record_translator::errc::translation_error:
         return o << "record_translator::errc::translation_error";
+    case record_translator::errc::unexpected_schema:
+        return o << "record_translator::errc::unexpected_schema";
+    }
+}
+
+record_type
+default_translator::build_type(std::optional<resolved_type> val_type) {
+    if (val_type.has_value()) {
+        return structured_translator.build_type(std::move(val_type));
+    }
+    return kv_translator.build_type(std::move(val_type));
+}
+
+ss::future<checked<iceberg::struct_value, record_translator::errc>>
+default_translator::translate_data(
+  model::partition_id pid,
+  kafka::offset o,
+  iobuf key,
+  const std::optional<resolved_type>& val_type,
+  iobuf parsable_val,
+  model::timestamp ts,
+  const chunked_vector<std::pair<std::optional<iobuf>, std::optional<iobuf>>>&
+    headers) {
+    if (val_type.has_value()) {
+        co_return co_await structured_translator.translate_data(
+          pid,
+          o,
+          std::move(key),
+          val_type,
+          std::move(parsable_val),
+          ts,
+          headers);
+    }
+    co_return co_await kv_translator.translate_data(
+      pid, o, std::move(key), val_type, std::move(parsable_val), ts, headers);
+}
+
+record_type key_value_translator::build_type(std::optional<resolved_type>) {
+    auto ret_type = schemaless_struct_type();
+    std::optional<schema_identifier> val_id;
+    auto& system_fields = std::get<iceberg::struct_type>(
+      ret_type.fields[0]->type);
+    system_fields.fields.emplace_back(iceberg::nested_field::create(
+      10, "data", iceberg::field_required::no, iceberg::binary_type{}));
+    return record_type{
+      .comps = record_schema_components{
+          .key_identifier = std::nullopt,
+          .val_identifier = std::move(val_id),
+      },
+      .type = std::move(ret_type),
+    };
+}
+
+ss::future<checked<iceberg::struct_value, record_translator::errc>>
+key_value_translator::translate_data(
+  model::partition_id pid,
+  kafka::offset o,
+  iobuf key,
+  const std::optional<resolved_type>& val_type,
+  iobuf parsable_val,
+  model::timestamp ts,
+  const chunked_vector<std::pair<std::optional<iobuf>, std::optional<iobuf>>>&
+    headers) {
+    if (val_type.has_value()) {
+        vlog(
+          datalake_log.error,
+          "Must not have parsed schema when using key-value mode");
+        co_return record_translator::errc::unexpected_schema;
     }
+    auto ret_data = iceberg::struct_value{};
+    auto system_data = build_rp_struct(pid, o, std::move(key), ts, headers);
+    system_data->fields.emplace_back(
+      iceberg::binary_value{std::move(parsable_val)});
```

Comment:
BTW the proposal and agreement from Matt and the schema's doc is that in kv mode, the value is not nested in Redpanda... There are pros and cons to either way, but it's awkward to have a single "column" in the kv mode.

We can fix in another PR since we already have a lot to unwind here in tests.

Reply:
Ah missed that, thanks for the callout